flock.go

```diff
@@ -39,6 +39,25 @@ func NewFlock(path string) *Flock {
 	return &Flock{path: path}
 }
 
+// Close will call Unlock() and close the underlying file descriptor. Use of this
+// function is not strictly necessary as the file descriptor will be cleaned up
+// by the garbage collector, but some long lived applications may not want to
+// depend on that
+func (f *Flock) Close() error {
+	f.m.Lock()
+	defer f.m.Unlock()
+
+	f.unlockInternal()
```

Comment:
How would you like to deal with it? It seems like the same type of thing as line 55.

Reply:
In this case the most reasonable is likely to annotate the error and return it upward.

Reply:
Hmm. Well, would that be right? What happens if someone closes after `Unlock()`? Does it make sense to make it so that `Close()` just fails if `Unlock()` hasn't been called yet?

Reply:
I've been doing some thinking about this, and from my perspective I think I see two possible ways to go about solving it. I'm coming at it from the perspective of wanting to know when unsafe data access may have been happening. If I hit an error when trying to release the exclusive lock I have on a file, that may indicate I've been silently corrupting data and may want to manually validate consistency.

So I think there're two ways we can make sure that information isn't lost:

- The first would be to simply return the error from `f.unlockInternal()` if one occurs. I don't recall whether calling `.Unlock()` and then `.Close()` would result in `f.unlockInternal()` returning an error for on `.Close()`. If it does...
- The second option is to disallow closing when the file is still locked. This would require consumers to be more thoughtful of how they consume the package, and to do the orchestration in the right order.

@virtuald Which path do you feel better aligns with the problem you're solving?

Reply:
I think disallowing closing when the file is still locked is unintuitive and would lead to difficult to diagnose issues (particularly if one did an unchecked `defer lock.Close()` which seems to be a common if ill-advised pattern). 

According to [fcntl(2)](http://man7.org/linux/man-pages/man2/fcntl.2.html):

> If a process closes any file descriptor referring to a file, then all of the process's locks on that file are released, regardless of the file descriptor(s) on which the locks were obtained.

According to [MSDN for LockFile](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-lockfile):

> If a process terminates with a portion of a file locked or closes a file that has outstanding locks, the locks are unlocked by the operating system. However, the time it takes for the operating system to unlock these locks depends upon available system resources. Therefore, it is recommended that your process explicitly unlock all files it has locked when it terminates. If this is not done, access to these files may be denied if the operating system has not yet unlocked them.

From both of these perspectives, I think it's a valid thing to just ignore the unlock error, as Close will also unlock the file. If any error should be returned, I would say the error returned from Close() should be returned as that's the semantic thing that's occurring.